src/base/message_builder.rs

```diff
@@ -262,6 +260,31 @@ impl<Target: Composer> MessageBuilder<Target> {
         Ok(builder.answer())
     }
 
+    /// Starts creating an error for the given message.
+    ///
+    /// Like [`start_answer()`] but infallible. Questions will be pushed if possible.
+    pub fn start_error<Octs: Octets + ?Sized>(
+        mut self,
+        msg: &Message<Octs>,
+        rcode: Rcode,
+    ) -> AnswerBuilder<Target> {
+        {
+            let header = self.header_mut();
+            header.set_id(msg.header().id());
+            header.set_qr(true);
+            header.set_opcode(msg.header().opcode());
+            header.set_rd(msg.header().rd());
+        }
+        let mut builder = self.question();
+        for item in msg.question().flatten() {
+            if builder.push(item).is_err() {
+                break;
```

Comment:
I think this is wrong. Under normal circumstances, a question will always fit. So if it doesn't then the rcode needs to be changed to reflect what is wrong. For example, if the first question doesn't fit then it should be SERVFAIL. If there is a second question, then FORMERR is more appropriate.

Reply:
I think I was thinking that if push() fails there are serious memory issues and at that point attempting to make any further changes may also fail, better to just stop and try to work with what we have. I have changed it to reset the question section to empty and set the rcode to SERVFAIL. Is that better?

Reply:
So don't reset the question section, just set rcode to SERVFAIL?

Reply:
Done.